src/main/java/com/znsio/teswiz/runner/Driver.java

```diff
@@ -603,4 +603,72 @@ public void doubleTap(WebElement element) {
                 .addAction(touch.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
         appiumDriver.perform(Arrays.asList(clickPosition));
     }
+
+    private Sequence fingerAction(String fingerName, Point locus, int startRadius, int endRadius, double angle, Duration duration) {
+
+        PointerInput finger = new PointerInput(PointerInput.Kind.TOUCH, fingerName);
+        Sequence fingerPath = new Sequence(finger, 0);
+
+        int fingerStartXPoint = (int) Math.floor(locus.x + startRadius * Math.cos(angle)); //converting from polar coordinates to cartesian
+        int fingerStartYPoint = (int) Math.floor(locus.y - startRadius * Math.sin(angle));
+
+        int fingerEndXPoint = (int) Math.floor(locus.x + endRadius * Math.cos(angle));
+        int fingerEndYPoint = (int) Math.floor(locus.y - endRadius * Math.sin(angle));
+
+        fingerPath.addAction(finger.createPointerMove(Duration.ZERO, PointerInput.Origin.viewport(), fingerStartXPoint, fingerStartYPoint))
+                .addAction(finger.createPointerDown(PointerInput.MouseButton.LEFT.asArg()))
+                .addAction(new Pause(finger, Duration.ofMillis(10)))
+                .addAction(finger.createPointerMove(duration, PointerInput.Origin.viewport(), fingerEndXPoint, fingerEndYPoint))
+                .addAction(finger.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
+
+        return fingerPath;
+    }
+
+    private Collection<Sequence> pinchAndZoom(Point locus, int startRadius, int endRadius, int pinchAngle, Duration duration) {
+
+        double angle = Math.PI / 2 - (2 * Math.PI / 360 * pinchAngle); // convert degree angle into radians
+
+        Sequence finger1Path = fingerAction("finger1", locus, startRadius, endRadius, angle, duration);
+
+        angle = angle + Math.PI;
+        Sequence finger2Path = fingerAction("finger2", locus, startRadius, endRadius, angle, duration);
+
+        return Arrays.asList(finger1Path, finger2Path);
+    }
+
+    private Collection<Sequence> pinchAndZoomIn(Point locus, int distance) {
+        int startRadius = 200, endRadius = 200 + distance, pinchAngle = 45, duration = 100;
+        return pinchAndZoom(locus, startRadius, endRadius, pinchAngle, Duration.ofMillis(duration));
+    }
+
+    private Collection<Sequence> pinchAndZoomOut(Point locus, int distance) {
+        int endRadius = 200, startRadius = 200 + distance, pinchAngle = 45, duration = 100;
+        return pinchAndZoom(locus, startRadius, endRadius, pinchAngle, Duration.ofMillis(duration));
+    }
+
+    public void pinchAndZoomIn(WebElement element) {
+
+        AppiumDriver appiumDriver = (AppiumDriver) this.driver;
```

Comment:
there are no log statements here that can assist in debugging

Reply:
Added debug logs